google/cloud/spanner/internal/connection_impl_test.cc

```diff
@@ -497,6 +500,69 @@ TEST(ConnectionImplTest, RollbackSuccess) {
   EXPECT_TRUE(rollback.ok());
 }
 
+TEST(ConnectionImplTest, MultipleThreads) {
+  auto db = Database("project", "instance", "database");
+  std::string const session_prefix = "test-session-prefix-";
+  std::string const transaction_id = "test-txn-id";
+  int session_counter = 0;
+
+  auto mock = std::make_shared<spanner_testing::MockSpannerStub>();
+  EXPECT_CALL(*mock, CreateSession(_, _))
+      .WillRepeatedly(
+          Invoke([&db, &session_prefix, &session_counter](
+                     grpc::ClientContext&,
+                     spanner_proto::CreateSessionRequest const& request) {
+            EXPECT_EQ(db.FullName(), request.database());
+            spanner_proto::Session session;
+            session.set_name(session_prefix +
+                             std::to_string(++session_counter));
+            return session;
+          }));
+  EXPECT_CALL(*mock, Rollback(_, _))
+      .WillRepeatedly(Invoke(
+          [session_prefix](grpc::ClientContext&,
+                           spanner_proto::RollbackRequest const& request) {
+            EXPECT_THAT(request.session(), StartsWith(session_prefix));
+            return Status();
+          }));
+
+  auto conn = std::make_shared<ConnectionImpl>(db, mock);
+
+  int const per_thread_iterations = 1000;
+  auto const thread_count = []() -> unsigned {
+    if (std::thread::hardware_concurrency() == 0) {
+      return 16;
+    }
+    return std::thread::hardware_concurrency();
+  }();
+
+  auto runner = [](int thread_id, int iterations,
+                   std::shared_ptr<Connection> const& conn) {
```

Comment:
Sure, but it has to be `T&` because we are calling non-const member functions.  Note that this requires a `std::ref` below because `std::async` copies every argument.